### PR TITLE
Payments Management: Show "Add Credit Card" CTA for full-credits purchase warning

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

The button on the warning for full-credits purchases should not show "Renew now" because the purchase may have just been renewed.

Due to the complexity of the logic in this component, it can be hard to reason about the effects of any change. To that end, this PR adds unit tests for many cases of the component's rendering. It doesn't cover all the cases because there are so many, but it does provide a mechanism to test the case of this issue and any case we like in the future, and any test we add will prevent regressions.

Originally reported by @nbloomf here: https://github.com/Automattic/wp-calypso/pull/46069#issuecomment-702327064

Fixes 9-gh-payments-shilling

Before:

Screen Shot 2021-02-03 at 1 30 36 PM

After:

Screen Shot 2021-02-03 at 1 31 05 PM

#### Testing instructions

- Visit `/me/purchases` and select a purchase recently made (or renewed) fully with credits.
- Make sure that you see the warning banner at the top of the subscription page.
- Make sure the button on the warning reads "Add Payment Method".
- Repeat the above steps for a purchase made fully with credits but that will expire soon. Verify that the button reads "Renew Now".

### CODE REVIEW

**Review by nbloomf — APPROVED** [review]
Works as expected! Leaving a note to myself to test this from the `purchases/subscriptions` route after this and #46333 have merged

**Review by fditrapani — APPROVED** [review]
Thanks for making this update. Works as expected!

**Review by DavidRothstein — CHANGES_REQUESTED** [review]
This doesn't look right to me, actually.  What happens if the purchase is close to expiration, or if it has already expired?  We don't want to show the "Add Credit Card" message then...

**Comment by DavidRothstein** [comment]
Below are before and after screenshots showing what I mean.  This is in the case of an expired subscription, where it's especially bad to tell people to add a credit card (because it might never actually be auto-renewed if the subscription is already expired).  But the issue exists for a subscription that is expiring soon also, since we want the "Renew Now" messaging to match what's used elsewhere on the page.

I'm not sure what the exact fix should be, but I think it involves trying to solve the bug introduced in #35723 at the same time (see internal discussion in 9-gh-payments-shilling) so as to rewrite this method completely so that it uses a different strategy for picking which message to display.

Before this pull request:

expired-before

After this pull request:

expired-after

**Comment by DavidRothstein** [comment]
I wonder if something like this would work:

```
if ( isExpired( purchase ) || purchase.expiryStatus === 'expiring' ) {
  // Render "Renew Now"
} else if ( ! hasPaymentMethod( purchase ) || isPaidWithCredits( purchase ) ) {
  // Render "Add Credit Card"
}
```

It's hard to tell for sure because there are so many different possible conditions, including monthly-vs-yearly subscriptions and all sorts of other things...

Ideally, there would be `canExplicitRenew()` and `isRechargeable()` checks somewhere in there too (if `canExplicitRenew()` is false it should never say "Renew Now"; if `isRechargeable()` is false the user should be encouraged to add a credit card when the subscription doesn't expire soon, which would probably cover the `isPaidWithCredits()` case also).  But if including those it might be necessary to change the message text in `getExpiringLaterText()` to match.

**Comment by sirbrillig** [comment]
> I wonder if something like this would work:

This is tricky because what do we render if both of those conditions are false? Right now the fallback is "Renew now", but it seems that the implicit conditions around that fallback are the problem. I suppose we could render no button in that case, but that's probably not the right behavior either.

I keep trying to untangle this class and it's _very_ hard. I suspect that, like the regression you noticed above, there are lots of places where the component relies on coincidence to render the correct thing. I think it might be wise for someone to eventually add unit tests to every possible case so that it can be refactored safely.

In the meantime, all we can really do is make our best guess.

(As an aside, it's interesting that last year this was refactored in https://github.com/Automattic/wp-calypso/pull/34076 and prior to that, `shouldAddPaymentSourceInsteadOfRenewingNow()` only returned true for credits purchases. I suspect this current bug is caused by that PR mixing with whatever change started using that function in this notice.)

I tried to go through the existing logic of `renderRenewNoticeAction()` and reduce it to set of all possible cases:

- If we have a payment method (including credits), render "Renew now".
- If we have no payment method, and we can't explicitly renew, render "Add credit card".
- If we have no payment method, but we can explicitly renew, and the expiry date is not within<sup>1</sup> than 3 months, render "Add credit card".
- If we have no payment method, but we can explicitly renew, and the expiry date is within<sup>1</sup> 3 months, render "Renew now".

We want to add to and alter these cases in this PR. Based on what you wrote above and the original issue, here's what I believe they are:

- If we have a payment method that's not credits, render "Renew now".
- If we have a payment method that's credits, and the purchase is neither expiring nor expired, render "Add credit card".
- If we have a payment method that's credits, and the purchase is expiring or expired, render "Renew now".
- If we have no payment method, and we can't explicitly renew, render "Add credit card".
- If we have no payment method, but we can explicitly renew, and the expiry date is not within<sup>1</sup> than 3 months, render "Add credit card".
- If we have no payment method, but we can explicitly renew, and the expiry date is within<sup>1</sup> 3 months, render "Renew now".

<sup>1</sup>: I read `moment( purchase.expiryDate ) > moment().add( 3, 'months' )` to mean "the expiry date is not within 3 months", but I have trouble parsing that so I may have gotten these backwards.

With those cases in mind, I'm trying a different approach.

**Review by wp-desktop — DISMISSED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@sirbrillig please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/926766). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

**Comment by sirbrillig** [comment]
Just in case anyone is waiting on something: this PR has been ready to review again for 13 days.

**Comment by fditrapani** [comment]
I was able to confirm the add credit card CTA for a product purchased on credits. Is there a way for me to test expired subscriptions that don't appear in my list?

**Comment by fditrapani** [comment]
Confirmed for expired plans:

image

**Review by fditrapani — APPROVED** [review]

**Comment by DavidRothstein** [comment]
Sorry, I'm not sure how I missed this was up for review again.

I'm willing to believe that this version **might** come pretty close to preserving existing behavior (except for the bug being fixed here) but the reason I suggested the simpler rewrite above was that the existing behavior doesn't really make any sense, so I'm not sure it's worth a lot of effort to preserve :smile: For example:
1. Why would we show "Renew Now" as the highlighted action for an expired subscription paid with credits, but _not_ show "Renew Now" for the much more common case of an expired subscription paid with a credit card?
2. Why would we ever show "Renew Now" as the highlighted action for something that isn't expiring soon (for example, something the user literally just bought 5 minutes ago)?  This is the kind of thing that contributes to people accidentally double-purchasing.
3. I'm not positive of this, but I'm suspicious of using `shouldAddPaymentSourceInsteadOfRenewingNow()` here at all, because I'm not sure it ever makes sense for a monthly subscription.

I do agree that it's also hard to trust my version is correct.  It's really hard to trust any proposed changes here!

>I keep trying to untangle this class and it's very hard. I suspect that, like the regression you noticed above, there are lots of places where the component relies on coincidence to render the correct thing. I think it might be wise for someone to eventually add unit tests to every possible case so that it can be refactored safely. In the meantime, all we can really do is make our best guess.

Or instead, just change the code to always define the message text and button in the same place (for example, change `getExpiringLaterText()` to return an object with the message text and button as properties).  Then it would be easy to understand what the user will see in each scenario without testing everything.  I'm not sure if that's something that should happen here or elsewhere, but it would make it much easier to have confidence that the messages and buttons always make sense together.

**Comment by sirbrillig** [comment]
> Or instead, just change the code to always define the message text and button in the same place. Then it would be easy to understand what the user will see in each scenario without testing everything. I'm not sure if that's something that should happen here or elsewhere, but it would make it much easier to have confidence that the messages and buttons always make sense together.

I agree; in fact I tried to do just that, but what I meant above is that such a refactor is quite complex and even though afterward we'd know what the output is, it's hard to know if we would be breaking some existing behavior. We'll need the manual testing regardless until we have automated testing.

> I suggested the simpler rewrite above was that the existing behavior doesn't really make any sense, so I'm not sure it's worth a lot of effort to preserve

You are probably right, but in some cases the behavior might be correct, and we might break it without some way to verify that it does what should be done.

My suggestion to add automated tests was not only to verify the existing behavior (which you're right might be buggy), but also to make that behavior explicit. That way if we write a test that says something that doesn't make sense, we can correct it easily (just by changing the test, which will make sure we change the associated code).

I should put my code where my keyboard is though, and start trying to add some such tests in this PR.

**Comment by sirbrillig** [comment]
The reason this had been de-prioritized was because of the complexity of the component which displays those messages; any change could potentially have far reaching consequences to other notices and it's hard to predict what will happen. If the component was covered in tests it would be much safer to make changes, but there are so many possible combinations of existing behavior (most which aren't necessary but we don't know which) that adding tests would take forever. This has stalled my ability to work on the change.

After discussion, we decided that there's risk no matter what we do, so it might be worth adding just a few tests to cover the behavior we care about and then making the change as best as we can. Then if there are regressions, we can fix those and also add tests for them.

**Comment by sirbrillig** [comment]
I've added unit tests for the simplest cases of this component, which covers expired and expiring purchases and should (I think) protect against the cases you mentioned in your initial review @DavidRothstein.

~Unfortunately, I think the behavior we actually want to change is in `renderOtherRenewablePurchasesNotice`, which is the only part I didn't test. (This actually makes me very curious how I got the screenshots in the PR description because AFAICT my changes would not have affected that function.)~ Nevermind. I forgot that `isExpiring()` returns true when autorenewal is disabled.

I think that the next step is to add tests for the specific behavior in `renderOtherRenewablePurchasesNotice` that we want to correct.

**Comment by sirbrillig** [comment]
With the added unit tests, I believe this should now cover the use cases we care about.

**Review by michaeldcain — APPROVED** [review]
This LGTM. Thanks for adding the test. I tested the following:

- If we have a payment method that's credits, and the purchase is neither expiring nor expired, render "Add Payment Method".
- If we have a payment method that's credits, and the purchase is expiring or expired, and the purchase is not rechargeable, render "Renew now".
- If we have no payment method, but we can explicitly renew, and the expiry date is within1 3 months, and the purchase is not rechargeable, render "Renew now".
- If we have no payment method, but we can explicitly renew, and the expiry date is not within1 than 3 months, render "Add Payment Method".
- Otherwise, render nothing.

**Comment by DavidRothstein** [comment]
Very belated review - this looks great, but there is still more we can fix based on the above discussion (it wasn't part of the original goal of this pull request), so I created a new issue for that: https://github.com/Automattic/wp-calypso/issues/50602

Also, as mentioned there, the tests added here are really great, but I think there might be some cases where they are not testing the correct thing (perhaps because of the way they set `isRechargeable: false`).  But that can probably be handled over there.

**Pull request merged**
